### ISSUE TITLE
No support for Feature & FeatureCollection in GeoJson

### ISSUE THREAD

**Issue by FreifeldRoyi** [locationtech/jts#419]
Hi
Feature & FeatureCollection is part of GeoJson spec for quite some time, and yet to be implemented in JTS (nor in Spatial4J).
Do you plan on releasing an implementation for it ?

**Comment by dr-jts** [comment]
The problem is that JTS has no notion of attributes.  So it's not clear how these would be provided to the GeoJSON I/O classes, or in fact whether to do this.

The GeoJSON IO classes were really provided to be embedded in other JSON frameworks which needed to handled geometry.

**Comment by FreifeldRoyi** [comment]
If I parse the Feature & FeatureCollection, will it be possible to hand the inner geometry type to JTS as a string and get a Geometry back, or should I construct everything manually? Or is there some other approach? 
I know that https://github.com/bjornharrtell/jts2geojson somewhat supports these features. Is there a better practice?

**Comment by dr-jts** [comment]
> If I parse the Feature & FeatureCollection, will it be possible to hand the inner geometry type to JTS as a string and get a Geometry back

Yes, that is the target use case for the GeoJSONReader.

> I know that https://github.com/bjornharrtell/jts2geojson somewhat supports these features. Is there a better practice?

I will check with the original author of these classes to see what his architecture was.

**Comment by dr-jts** [comment]
The GeoJSONReader class is designed to be used in the context of JSON frameworks which allow custom deserializers to be defined.  E.g. the [Jackson](https://github.com/FasterXML/jackson) API.

Here's a (partial) code snippet which shows how it can be used in Jackson:

```java
import org.codehaus.jackson.JsonNode;
import org.codehaus.jackson.JsonParser;
import org.codehaus.jackson.JsonProcessingException;
import org.codehaus.jackson.ObjectCodec;
import org.codehaus.jackson.map.DeserializationContext;
import org.codehaus.jackson.map.JsonDeserializer;

public class GeoJsonJacksonDeserializer extends JsonDeserializer<Geometry>
{
	public Geometry deserialize(JsonParser jsonParser, DeserializationContext context)
	{
		ObjectCodec oc = jsonParser.getCodec();
		JsonNode node = oc.readTree(jsonParser);
		String geoJson = node.toString();
		GeoJsonReader geoJsonReader = new GeoJsonReader();

		Reader reader = new StringReader(geoJson);
		Geometry result = geoJsonReader.read(reader);
```

**Comment by DontTrustMicrosoft-MicrosoftIsAsshole** [comment]
So after weeks of moving my project to GeoJson format from WKT, which open street map exports directly to, I have to learn that you wrote a GeoJsonReader that doesn't understand perfectly fine formated GeoJson strings? (https://en.wikipedia.org/wiki/GeoJSON)  Am I that sleep deprevated that I fail to understand the logic? I'm a bit speechless here
-> your WKT reader understands and reads WKT strings
-> your GeoJsonReader doesn't understand or read GeoJson string
Why not implement the RFC standard in your GeoJson code and constants, so we as end users don't have to worry about coding around it? I now have to learn how to correctly "extract" polygones, multipolygones, geometries and all that out of a GeoJson file. What do I need you libraries for then?

**Comment by danishnawab** [comment]
@dr-jts Would it be a fair compromise to reduce GeoJSON's `Feature` to JTS' `Geometry` (and consequently `FeatureCollection` to `GeometryCollection`)? 
Since JTS doesn't have a notion of additional `properties` (or `id`), could we just jettison those attributes and still map the standard GeoJSON types to the existing (and compatible) JTS types?
The GeoJSON RFC doesn't ascribe any special meaning to the `properties` object, so I suppose there is no harm in dropping it. 
I could submit a PR if this proposal resonates with you.

**Comment by dr-jts** [comment]
@danishnawab yes, that seems like a reasonable extension.  A PR for review is welcome.

**Comment by danishnawab** [comment]
@dr-jts Here's the PR: https://github.com/locationtech/jts/pull/837
